### PR TITLE
Add permutation and space sequence spaces

### PR DESCRIPTION
**Pull request by sogartar**
This PR depends on https://github.com/facebookresearch/CompilerGym/pull/643. It has to be merged after it. During review please ignore the first commit and review only the ones after that.

It also assumes that https://github.com/facebookresearch/CompilerGym/pull/644 will be merged before that to have consistency in space comparison. If the space name should not be considered during comparison I will remove it from this PR as well.

### CODE REVIEW

File: compiler_gym/service/proto/CMakeLists.txt
Comment (ChrisCummins) [inline compiler_gym/service/proto/CMakeLists.txt]: Is this whitespace-only change generated by the pre-commit hook? Is it needed?

File: compiler_gym/service/proto/compiler_gym_service.proto
Comment (ChrisCummins) [inline compiler_gym/service/proto/compiler_gym_service.proto:298]: This new proto needs to be added to https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/rpc.rst#request-and-reply-messages

File: compiler_gym/service/proto/py_converters.py
Comment (ChrisCummins) [inline compiler_gym/service/proto/py_converters.py]: Could this error message be improved to specify what the failed requirement is?

File: compiler_gym/spaces/permutation.py
Comment (ChrisCummins) [inline compiler_gym/spaces/permutation.py:13]: Needs a docstring and entry in https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/compiler_gym/spaces.rst

File: compiler_gym/spaces/space_sequence.py
Comment (ChrisCummins) [inline compiler_gym/spaces/space_sequence.py:14]: Same as above, needs documentation

File: compiler_gym/spaces/permutation.py
Comment (ChrisCummins) [inline compiler_gym/spaces/permutation.py:16]: The constructor needs a docstring and `   .. automethod:: __init__` in the rst

File: compiler_gym/spaces/space_sequence.py
Comment (ChrisCummins) [inline compiler_gym/spaces/space_sequence.py]: Why declare these here?

File: tests/util/permutation_test.py
Comment (ChrisCummins) [inline tests/util/permutation_test.py:1]: No test coverage of invalid size range

File: tests/util/permutation_test.py
Comment (ChrisCummins) [inline tests/util/permutation_test.py:17]: Nitpick: please use `np.testing.assert_array_equal(...)` instead of `assert np.array_equal(...)` as it produces better error messages on failure.

**Review by ChrisCummins — APPROVED** [review]
LGTM, with a few niggles

File: compiler_gym/service/proto/CMakeLists.txt
Comment (sogartar) [reply compiler_gym/service/proto/CMakeLists.txt]: I am not sure why it changed. I was rebasing after the introduction of the cmake hook. I removed these changes.

File: compiler_gym/service/proto/compiler_gym_service.proto
Comment (sogartar) [reply compiler_gym/service/proto/compiler_gym_service.proto:298]: Done.

File: compiler_gym/service/proto/py_converters.py
Comment (sogartar) [reply compiler_gym/service/proto/py_converters.py]: Done.

File: compiler_gym/spaces/permutation.py
Comment (sogartar) [reply compiler_gym/spaces/permutation.py:13]: Done.

File: compiler_gym/spaces/space_sequence.py
Comment (sogartar) [reply compiler_gym/spaces/space_sequence.py:14]: Done.

File: compiler_gym/spaces/permutation.py
Comment (sogartar) [reply compiler_gym/spaces/permutation.py:16]: Done.

File: compiler_gym/spaces/space_sequence.py
Comment (sogartar) [reply compiler_gym/spaces/space_sequence.py]: I removed them.

File: tests/util/permutation_test.py
Comment (sogartar) [reply tests/util/permutation_test.py:1]: I added a test and also for the conversion from proto.

File: tests/util/permutation_test.py
Comment (sogartar) [reply tests/util/permutation_test.py:17]: Done.

File: tests/util/permutation_test.py
Comment (ChrisCummins) [reply tests/util/permutation_test.py:1]: Thanks!

**Review by ChrisCummins — APPROVED** [review]
LGTM, thanks for quickly addressing my comments!

**Review by ChrisCummins — APPROVED** [review]

**Comment by ChrisCummins** [comment]
@sogartar, there is a merge conflict, feel free to merge yourself after address, or you can ping me to merge.

EDIT: nvm, it was a trivial fix 🙂 

Cheers,
Chris

**Comment by sogartar** [comment]
@ChrisCummins, are you going to merge this PR?

**Comment by ChrisCummins** [comment]
Ah, sorry for the delay! I must have forgot to press the merge button 🙂

**Pull request merged**
